### ISSUE TITLE
关于FFT_for_Period中frequency_list和top_list计算方式

### ISSUE THREAD

**Issue by gengdd** [thuml/Time-Series-Library#62]
作者您好，今天在阅读TimesNet代码时发现`FFT_for_Period`中存在Batch内部数据之间的交互。代码片段如下：

```python
# https://zhuanlan.zhihu.com/p/607836445
# 先把xf全部都取绝对值变非负，再在把频域中的信号在batch、channel中取均值
frequency_list = abs(xf).mean(0).mean(-1)
frequency_list[0] = 0 
_, top_list = torch.topk(frequency_list, k)
```
如此设计的好处是明显的，一个Batch内部使用了相同的周期长度，`reshape`二维时会很方便。但是请问此种操作是否存在以下两个潜在的问题：

1. 数据泄露。如果数据没有进行`shuffle`，Batch内部的数据形式大致如下： $[[x_1, x_2, x_3],[x_2, x_3, x_4],\cdots]$ ，即同一个Batch内部下边的数据feature中存在了上边数据的预测值label，这样在做`abs(xf).mean(0)`是否会数据泄露。简而言之，如果Batch内部样本特征存在了其他样本的真实值，进行Batch样本间计算是否属于数据泄露。
2. 同一个样本预测结果不一致。由于同一Batch内部进行了`mean`计算，是否意味着如果同一个样本如果在不同Batch中计算`mean`结果会不一样，从而导致预测值不唯一。作者是否做过如下测试：首先单样本输入模型进行预测，然后将这个单样本和其他样本形成不同的Batch数据输入模型进行预测，这个单样本的预测值是否会出现不一致的现象。

**Comment by wuhaixu2016** [comment]
您好，感谢您的细致阅读～
（1）目前时间序列预测在evaluation阶段都是按照batch size=1进行检验的，所以没有数据泄漏。
（2）在训练阶段，batch norm也是会随着数据变化而结果变化，但是因为在测试阶段，batch size=1，所以预测结果也不会有变化。

**Comment by gengdd** [comment]
> 您好，感谢您的细致阅读～ （1）目前时间序列预测在evaluation阶段都是按照batch size=1进行检验的，所以没有数据泄漏。 （2）在训练阶段，batch norm也是会随着数据变化而结果变化，但是因为在测试阶段batch size=1，所以预测结果也不会有变化。
> 
> 另外，在实际应用中，模型部署之后都是不断滚动预测的，所以eval阶段的batch size设置为1也是符合现实场景要求的。


感谢作者回复，因为今天没跑代码，所以没了解到evaluation阶段Batch_size=1的设置。另外，个人不太确定此种计算是否能和Batch Norm同等类比，个人理解BatchNorm在测试阶段也不会受到Batch size大小的影响，而此种方式计算同一样本的预测值会受到Batch size的影响，当然Batch size=1符合现实场景的解释是合理的，那么请问如何Batch size不为1，同一样本是否存在预测值不一致的可能？

**Comment by wuhaixu2016** [comment]
这个也是不存在的，因为在测试阶段，数据之间没有随机性，只要batch size固定，模型的输出结构就是确定的。

这里其实涉及一个trade-off：（1）如果多个时间序列一起计算周期，那么可以排除一些噪音的影响，对于周期的估计会更加准确；（2）如果每个时间序列分别计算周期，那么对于每个序列的特异性信息会处理地更好，但是可能在寻找周期长度上会遇到一些干扰。

**Comment by gengdd** [comment]
> 

感谢作者的耐心回复，确实如您所说如此设计模型性能更好更加鲁棒。但是我想表达的是如下疑问：
```python

# 同一份dataset
dataset = xxx

# 设置不同的batch size大小
dataloader1 = DataLoader(dataset, batch_size=32)
dataloader2 = DataLoader(dataset, batch_size=128)

y_pred1 = []
for batch in dataloader1:
    pred = model(batch)
    y_pred1.append(pred)
y_pred1 = torch.stack(y_pred1)

y_pred2 = []
for batch in dataloader2:
    pred = model(batch)
    y_pred2.append(pred)
y_pred2 = torch.stack(y_pred2)

y_pred1 == y_pred2   # ???是否相等，如此设计是否会造成不相等？
```
我想咨询的是不同`Batch size`的同一样本的预测值是否相等。如果不相等，是否关键？因为我之前的认为是同一样本输入，模型输出应该保持一致。同一样本在不同`Batch size`设置下输入，模型输出不一致的情况是否之前存在过，或者有被讨论过。

**Comment by wuhaixu2016** [comment]
您好，如果严格来说，确实不一定完全相等，但是在batch维度求mean只是用于挑选top k的周期，这个数值对于某一个数据集来说基本上变化不大，我们实验也是大部分的实验结果是相同的。
如果不相同的话，我觉得影响不大，还是那个场景，训练的时候batch size不为1，输出结果不一样，但是仅仅是影响优化；测试过程batch size为1，不会影响输出结果。

**Comment by gengdd** [comment]
> 您好，如果严格来说，确实不一定完全相等，但是在batch维度求mean只是用于挑选top k的周期，这个数值对于某一个数据集来说基本上变化不大，我们实验也是大部分的实验结果是相同的。 如果不相同的话，我觉得影响不大，还是那个场景，训练的时候batch size不为1，输出结果不一样，但是仅仅是影响优化；测试过程batch size为1，不会影响输出结果。

感谢回复。在训练期间所造成的样本预测值不一致问题是问题不大的，问题就在于在模型测试阶段样本预测值会受到的Batch sized大小的影响，这就与Batch Norm不一样了。当然，在时序预测场景下可能是没问题的，因为Batch size为1的设定正好规避了此种设计的缺点，同时享受了其优点。感谢作者的答疑解惑！

**Comment by gengdd** [comment]
> 您好，如果严格来说，确实不一定完全相等，但是在batch维度求mean只是用于挑选top k的周期，这个数值对于某一个数据集来说基本上变化不大，我们实验也是大部分的实验结果是相同的。 如果不相同的话，我觉得影响不大，还是那个场景，训练的时候batch size不为1，输出结果不一样，但是仅仅是影响优化；测试过程batch size为1，不会影响输出结果。

感谢回复。在训练期间所造成的样本预测值不一致问题是问题不大的，问题就在于在模型测试阶段样本预测值会受到的Batch sized大小的影响，这就与Batch Norm不一样了。当然，在时序预测场景下可能是没问题的，因为Batch size为1的设定正好规避了此种设计的缺点，同时享受了其优点。感谢作者的答疑解惑！
